userdocs/diagnostics/temporary-pointers.md

```diff
@@ -0,0 +1,36 @@
+# Temporary Pointers
+A temporary, or ephemeral, pointer in Swift is a pointer which is introduced by an implicit function argument conversion and is only valid for the lifetime of the function call it appears in. There are a few ways to create a temporary pointer:
+
+- Using an inout-to-pointer conversion:
+
+  ```swift
+  func foo(bar: UnsafePointer<Int>) { /*...*/ }
+  var x: Int = 42
+  foo(bar: &x)
+  ```
+
+  In the example above, the `bar` passed to `foo` is a temporary pointer to `x` which is only valid until `foo` returns. It's important to note that not all inout-to-pointer conversions create a temporary pointer. Global variables and stored properties thereof can be used as non-ephemeral pointers when passed inout.
+
+- Using a string-to-pointer conversion:
+
+  ```swift
+  func foo(bar: UnsafePointer<Int8>) { /*...*/ }
+  var x: String = "hello, world!"
+  foo(bar: x)
+  ```
+
+  In the example above, the `bar` passed to `foo` is a temporary pointer to a buffer containing the UTF-8 code units of `x` which is only valid until `foo` returns.
+
+- Using an array-to-pointer conversion:
+
+  ```swift
+  func foo(bar: UnsafePointer<Bool>) { /*...*/ }
+  var x: [Bool] = [true, false, true]
+  foo(bar: x)
+  ```
+
+  In the example above, the `bar` passed to `foo` is a temporary pointer to the contents of `x` which is only valid until `foo` returns.
+
+Temporary pointers may only be passed as arguments to functions which do not store the pointer value or otherwise allow it to escape the function's scope. The Swift compiler is able to diagnose some, but not all, violations of this rule. Misusing a temporary pointer by allowing it to outlive the enclosing function call results in undefined behavior.
+
+When calling functions which cannot safely accept temporary pointer arguments, use the `withUnsafe*` family of pointer APIs to explicitly manage pointer lifetimes.
```

Comment:
This for doing this. I think it's great and agree with all the feedback so far.

> When calling functions which cannot safely accept temporary pointer arguments, use the `withUnsafe*` family of pointer APIs to explicitly manage pointer lifetimes.

Mentioning `withUnsafePointer` opens up a dangerous door. I think it's the biggest issue with the current warnings. The usual reaction will be to think that merely using `withUnsafePointer` imbues code with some extra safety--of course it doesn't, it just silences warnings. The educational note is the perfect place to clear this up.

If we leave it as "use withUnsafePointer to manage lifetime" without explaining how to "manage lifetimes", we'll just get a lot of this sort of thing:

```
struct PointerWrapper {
  var ptr: UnsafePointer<Int>
}

func foo(x: Int) -> PointerWrapper {
  return withUnsafePointer(to: x) {
    PointerWrapper(ptr: $0)
  }
}
```

It may help just to search through some common Swift packages to see common cases of programmers using withUnsafePointer correctly, and contrast that with incorrect usage.